control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go

```diff
@@ -456,6 +389,105 @@ func (r *Reconciler) reconcileInPlaceUpgradeManifests(ctx context.Context, hoste
 	} else {
 		log.Info("Reconciled ConfigMap", "result", result)
 	}
+
+	// We will deploy the machine-config-daemon daemonset here, so it can be running for future operations
+	daemonset := inPlaceUpgradeDaemonset(namespace.Name)
+	if result, err := r.CreateOrUpdate(ctx, hostedClusterClient, daemonset, func() error {
+		return r.reconcileDaemonset(
+			daemonset,
+			poolName,
+			mcoImage,
+		)
+	}); err != nil {
+		return fmt.Errorf("failed to reconcile upgrade daemonset: %w", err)
+	} else {
+		log.Info("Reconciled upgrade daemonset", "result", result)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcileDaemonset(daemonset *appsv1.DaemonSet, poolName, mcoImage string) error {
+	configmap := inPlaceUpgradeConfigMap(poolName, daemonset.Namespace)
+	hostPathType := corev1.HostPathUnset
+	daemonset.Spec.Template.ObjectMeta = metav1.ObjectMeta{
+		Name: "machine-config-daemon",
+		Labels: map[string]string{
+			"k8s-app": "machine-config-daemon",
+		},
+	}
+	daemonset.Spec.Template.Spec.Containers = []corev1.Container{
+		{
+			Name:  "machine-config-daemon",
+			Image: mcoImage,
+			Command: []string{
+				"/usr/bin/machine-config-daemon",
+			},
+			Args: []string{
+				"start",
+				"--root-mount=/rootfs",
+				"--kubeconfig=/var/lib/kubelet/kubeconfig",
+				"--desired-configmap=/etc/machine-config-daemon-desired-config",
+			},
+			Env: []corev1.EnvVar{
+				{
+					Name: "NODE_NAME",
+					ValueFrom: &corev1.EnvVarSource{
+						FieldRef: &corev1.ObjectFieldSelector{
+							FieldPath: "spec.nodeName",
+						},
+					},
+				},
+			},
+			TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+			SecurityContext: &corev1.SecurityContext{
+				Privileged: k8sutilspointer.BoolPtr(true),
+			},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      "rootfs",
+					MountPath: "/rootfs",
+				},
+				{
+					Name:      "desired-config-mount",
+					MountPath: "/rootfs/etc/machine-config-daemon-desired-config",
+				},
+			},
+		},
+	}
+	daemonset.Spec.Template.Spec.HostNetwork = true
+	daemonset.Spec.Template.Spec.HostPID = true
+	daemonset.Spec.Template.Spec.Tolerations = []corev1.Toleration{
+		{
+			Operator: corev1.TolerationOpExists,
+		},
+	}
+	daemonset.Spec.Template.Spec.ServiceAccountName = "machine-config-daemon"
+	daemonset.Spec.Template.Spec.NodeSelector = map[string]string{
+		hyperv1.NodePoolLabel: poolName,
```

Comment:
Does this mean the pod is running idle  in every Node for the NodePool until the Node is actually chosen to upgrade?

Reply:
The daemonSet pod will actually get rescheduled even after the Node is upgraded until the whole NodePool is done and so the daemonSet deleted, right? I could think of envs with large NodePools that are "perpetually" upgrading where this means the pod is effectively long running. Could we somehow narrow scheduling to only the Nodes performing the upgrade within the NodePool?

Reply:
Yes, that's definitely an optimisation we should try to make. I will experiment with more dynamic labelling to select only nodes that require the daemonset.

This isn't by itself worse than today, since the `pod` we dynamically schedule also doesn't yet have a cleanup process until the whole update completes. If its ok with you, I'd like to make that a separate card/PR, since I would want to do some extra testing to ensure we are not missing some cases where a node needs action but the controller doesn't schedule the pod